### PR TITLE
Integrating Webpack 4

### PR DESCRIPTION
**Pull request by dzearing**
This change updates to webpack 4 and ts-loader.

It also adjusts `typings` locations in package.json, while also adding `module` entry points and `sideEffects: false` for tree shaking to work.

It does *not* change the `/lib/` references anywhere. We'll need to figure that out in a separate step.

### CODE REVIEW

File: scripts/package.json [hunk 1]
```diff
@@ -13,7 +13,6 @@
     "@microsoft/loader-load-themed-styles": "^1.6.0",
     "@uifabric/merge-styles": ">=5.13.0 <6.0.0",
     "autoprefixer": "^7.1.5",
-    "awesome-typescript-loader": "^3.2.3",
```
Comment (christiango) [inline scripts/package.json]: Why the switch?

File: scripts/tasks/webpack-resources.js [hunk 2]
```diff
@@ -52,13 +53,20 @@ module.exports = {
...
+        mode: 'production',
         output: {
           filename: `[name].min.js`,
           path: path.resolve(process.cwd(), 'dist')
         },
+
+        optimization: {
+          concatenateModules: true,
+          minimize: true
+        },
```
Comment (christiango) [inline scripts/tasks/webpack-resources.js]: these are the defaults for mode:production, you can omit this

Comment (christiango) [reply scripts/tasks/webpack-resources.js]: If you aren't seeing dev mode work, you probably need to update merge to make sure that you are deep copying all properties in the config when merging. We had a similar issue where our "merge" ended up with a reference to the same optimization object for development and production mode, which caused the defaults webpack applies in the devmode run to get applied in production mode.

Comment (dzearing) [reply scripts/tasks/webpack-resources.js]: Good catch, I know about `mode` but i was testing these individually,.

File: scripts/package.json (hunk 1 above)
Comment (dzearing) [reply scripts/package.json]: i'm going to revert to trim down changes.

**Pull request merged**